skrub/_reporting/_table_report.py

```diff
@@ -197,6 +198,30 @@ def _repr_mimebundle_(self, include=None, exclude=None):
     def _repr_html_(self):
         return self._repr_mimebundle_()["text/html"]
 
+    def write_html(self, filename):
```

Comment:
This looks great! In the issue I forgot to mention a couple of things:

- Sometimes generating the report can fail (unfortunately!). We want to make sure we close the file (if we opened it) in this case; also we may want to avoid truncating or creating it before we know we have something to write.
- The report will always set the charset to utf-8 in the page's metadata. So if the user passes an open file we may optionally check the file's encoding if we want to be extra cautious. Also note unless the doctring we write forbids it the user could pass a binary stream so we may want to handle that case, too.
- I'm not sure we need to enforce the filename suffix. I could imagine someone wanting to save to "report.htm" or "report.html.tmp" or something so I'd rather let the user name the file however they want.
- If we allow both a path and a file object maybe the parameter could be just "file" instead of "filename"?

So taking those details into account I guess the function could be modified slightly to look similar to this:

```python
    def write_html(self, file):
        html = self.html()
        if isinstance(file, (str, Path)):
            with open(file, "w", encoding="utf8") as stream:
                stream.write(html)
            return
        try:
            file.write(html.encode("utf-8"))
            return
        except TypeError:
            pass
        if (encoding := getattr(file, "encoding", None)) is not None:
            try:
                assert codecs.lookup(encoding).name == "utf-8"
            except (AssertionError, LookupError):
                raise ValueError(
                    f"If `file` is a text file it should use utf-8 encoding; got: {encoding!r}"
                )
        file.write(html)
```

Reply:
thank you comments @jeromedockes , I have couple of questions. 
1. What are the reasons behind `TypeError` 
2. if the encoding is None it seems the locale.encoding will be used, so we want to check for that too and raise an error if it is not utf-8 ?

Reply:
Hi

> 1. What are the reasons behind TypeError

Trying to write bytes and catching the `TypeError` is an easy way to detect if we are dealing with a text stream or a binary stream.

<details>

A file object (a.k.a. stream) can be:

- In binary mode, in which case it expects bytes that can be written to a file without transformation. If we want to write text we must encode it into bytes ourselves. Passing a `str` to `write` raises a `TypeError`.
- In text mode, in which case it expects strings and it will transparently encode them into bytes (and optionally translate newline characters) before writing the resulting bytes to the underlying binary buffer. Passing a `bytes` to `write` raises a `TypeError` (the one we catch).

```python
>>> 'é'
'é'
>>> 'é'.encode('utf-8')
b'\xc3\xa9'
```

Binary mode:

```python
>>> binary_stream = open('thefile', 'wb') # 'wb' = 'write + binary'
>>> binary_stream.write('é'.encode('utf-8')) # we can write bytes
2
>>> binary_stream.write(b'\xc3\xa9')
2
>>> binary_stream.write('é') # trying to write a string directly is a TypeError
Traceback (most recent call last):
    ...
TypeError: a bytes-like object is required, not 'str'
>>> binary_stream.close()
```

Conversely, if we use text mode (the default):

```python
>>> text_stream = open('thefile', 'wt', encoding='utf-8') # 'write + text', or 'w' as text is the default
>>> text_stream.write('é') # we can write strings
1
>>> text_stream.write(b'\xc3\xa9'.decode('utf-8'))
1
```

and this is the error we catch in the current version of `TableReport.write_html`:

```python
>>> text_stream.write(b'\xc3\xa9') # trying to write bytes to a text stream is a TypeError
Traceback (most recent call last):
    ...
TypeError: write() argument must be str, not bytes
>>> text_stream.close()
```

So we need to know if the file object the user gave us is a text or binary stream. If we knew it was created with `open` we could inspect its mode:

```python
>>> with open('thefile', 'w') as stream:
...    print(stream.mode)
w

>>> with open('thefile', 'wb') as stream:
...    print(stream.mode)
wb
```

But that may not always work with other kinds of file objects:

```python
>>> import io
>>> stream = io.BytesIO()
>>> stream.write(b'\xc3\xa9')
2
>>> stream.mode
Traceback (most recent call last):
    ...
AttributeError: '_io.BytesIO' object has no attribute 'mode'
```

All well-behaved file objects however will respect the rule of raising a TypeError if we give them the wrong type of data. So what we can do is first suppose it is in binary mode and try to write bytes, and if we get a TypeError try again with a string (or we could do it the other way around, too).

</details>


> 2. if the encoding is None it seems the locale.encoding will be used, so we want to check for that too and raise an error if it is not utf-8 ?

If we pass `encoding=None` to `open` it will default to using `locale.encoding`, but AFAIK the resulting file object will always have an explicit encoding (not `None`).

```python
>>> with open('thefile', 'w', encoding=None) as f:
...     print(f.encoding)
UTF-8
```

So I think it should be ok to inspect the encoding of the file object in this case.

<details>

Some other kinds of text stream may not have an explicit encoding though. For example `io.StringIO` writes to an in-memory python string, so it accepts strings but does not need to encode them (it doesn't need to convert to bytes because the backing storage is a unicode string)

```python
>>> stream = io.StringIO()
>>> stream.encoding is None
True
>>> stream.write('é') # ok
1
```

In the case of `StringIO` it's fine to copy the report string to the stream's buffer even if we couldn't figure out what encoding it uses (because it doesn't use any). In some other cases I guess looking at the `encoding` attribute may not give a result even if a (possibly wrong) encoding is being used. But I guess at this point we have to trust the user to know what they are doing and at least we have made a reasonable effort to prevent the most likely issue, which is doing this on windows:

```python
>>> with open('thefile', 'w') as f:
...     report.write_html(f)
```

(Even that will go away in python 3.15 when utf-8 becomes the default).
So my suggestion is that if we easily detect a wrong encoding is being used by checking the `encoding` attribute, we raise an error because we are sure that writing will result in a file with an incorrect `<meta charset='utf-8'>` tag which is likely to cause it to display incorrectly in a browser. If we cannot figure out for sure that a wrong encoding was used we just go ahead and write the report and hope for the best.

</details>

sorry for the long answer, hope this makes sense!

Reply:
I fail to see the use case for allowing the user to pass a file instead of a filename. According to some path, this feature should address the simple use case of writing an HTML file to disk. Can you motivate this a little bit? I looks like plain over-engineering to me TBH.

Reply:
I believe it was a suggestion from @glemaitre .

but if you look at functions offering to save some object as text such as matplotlib.savefig, pandas or polars to_csv etc, xml.etree.write and others it's hard to find an example that does not allow passing a file object.

one example use case would be using a [temporary file](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile), testing, situations in which you may need to write either to an actual file or an in-memory buffer, ...

from personal experience a while ago in a completely unrelated context NiftiImages from the nibabel library could only be saved to a filesystem path and I can't remember what were the exact situation where it was a problem for me but I remember it was annoying

Reply:
Thank you @jeromedockes for the explanation. it helped a lot.